### ISSUE TITLE
better pluggable reactor support for twistd

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10458]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#719|
|Type|enhancement|
|Created|2004-09-16 22:43:46Z|

<details><summary>Searchable metadata</summary>

```
trac-id__719 719
type__enhancement enhancement
reporter__itamarst itamarst
priority__highest highest
milestone__Core_2_5 Core-2.5
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1095374626000000 1095374626000000
changetime__1163557566000000 1163557566000000
version__ 
owner__ 
cc__exarkun cc__itamarst
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Currently I *think* adding new reactors just means adding entry to dict. If this
is sufficient we just need to document it (or maybe add tiny utility function
and  document that).

JP prefers to use .tml files for this, though.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
I'll do this with the new plugin system soon.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun removed owner|
|-|-|

Ready for review in pluggable-reactors-719-2

Most of the functionality added in this branch is in twisted.application.reactors: a new interface, IReactorInstaller, is added for reactor plugins, along with a helper class, Reactor, and some utility functions.

A bug is fixed in twistd which prevented a custom reactor from being selected when any subcommand plugin imported the reactor module (thus implicitly installing it before twistd got a chance to do so).

Some zsh completion functionality is dropped because it:

  * Had no test coverage
  * Made fixing twistd harder
  * I don't use zsh

If someone cares it could most likely be re-added without _too_ much difficulty, although tests will need to be added this time around.

I've added this to the 2.5 milestone since the Qt tickets already in the milestone depend on this.  Also, the twistd regression is pretty serious and should be fixed for 2.5.

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|PenguinOfDoom commented|
|-|-|

setUp/tearDown on PluggableReactorTestCase and TestRunner need docstrings.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set owner to @jml<br>@jml set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_7"></a>|@jml set owner to @exarkun<br>@jml set status to `new`|
|-|-|

Looks good. The tests in test_application are particularly apposite.

 * 80 cols in twisted/plugins/twisted_reactors
 * The trial manpage has options in alphabetic order. Please move --help-reactors to just before --help-reporters.

Please fix these then merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set status to `closed`|
|-|-|

(In [18675]) Merge pluggable-reactors-719-2

Author: exarkun, itamar
Reviewer: jml
Fixes [#10458](https://github.com/twisted/twisted/issues/10458)

Replace reactor dictionary with plugin-based definition mechanism.
Update twistd and trial to use the plugin system to list and install
reactors.

Also fix twistd so that alternate reactors can actually be selected
when using subcommands.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [18680]) Merge pluggable-reactors-719-2

Author: exarkun, itamar
Reviewer: jml, itamar (not the parts he wrote)
Fixes [#10458](https://github.com/twisted/twisted/issues/10458)

Replace reactor dictionary with plugin-based definition mechanism.
Update twistd and trial to use the plugin system to list and install
reactors.

Also fix twistd so that alternate reactors can actually be selected
when using subcommands.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
